example_test.go

```diff
@@ -180,6 +181,77 @@ Nested:
 	//    - B: 3
 }
 
+func ExampleAttributeEncoder_encode() {
+	// Create a netlink.AttributeEncoder that encodes to the same message
+	// as that decoded by the netlink.AttributeDecoder example.
+	ae, err := netlink.NewAttributeEncoder()
+	if err != nil {
+		log.Fatalf("failed to create attribute encoder: %v", err)
+	}
+
+	// nested is a nestes structure within out.
+	type nested struct {
+		A, B uint32
+	}
+
+	// out is an example structure we will use to pack netlink attributes.
+	type out struct {
+		Number uint16
+		String string
+		Nested nested
+	}
+
+	o := out{
+		Number: 1,
+		String: "hello world",
+		Nested: nested{
+			A: 2,
+			B: 3,
+		},
+	}
+
+	// encodeNested is an example function used to adapt the ae.Do method
+	// to encode an arbitrary structure.
+	//
+	// It is recommended to create functions such as this one at the package
+	// level, but for this example, it is necessary to create a closure to
+	// fully show the usage of netlink.AttributeEncoder.
+	encodeNested := func(n *nested) func() ([]byte, error) {
+		return func() ([]byte, error) {
+			// Create an internal, nested netlink.NewAttributeEncoder that
+			// operates on the nested set of attributes.
+			ae, err := netlink.NewAttributeEncoder()
+			if err != nil {
+				log.Fatalf("failed to create nested attribute encoder: %v", err)
+			}
+
+			// Encode the fields of the nested stucture
+			ae.Uint32(1, n.A)
+			ae.Uint32(2, n.B)
+
+			// Return the encoded attributes, and any error encountered.
+			return ae.MarshalBinary()
+		}
+	}
+
+	// Encode the Number attribute as a uint16.
+	ae.Uint16(1, o.Number)
+	// Encode the String attribute as a string.
+	ae.String(2, o.String)
+	// Nested is a nested structure, so we will use our encodeNested
+	// function along with ae.Do to encode it in a concise way.
+	ae.Do(3, encodeNested(&o.Nested))
+
+	b, err := ae.MarshalBinary()
+	// Any errors encountered during encoding (including any errors from
+	// encoding nested attributes) will be returned here.
+	if err != nil {
+		log.Fatalf("failed to encode attributes: %v", err)
+	}
+
+	fmt.Printf("Encoded netlink message follows:\n%s", hex.Dump(b))
```

Comment:
It'd be nice to have an `// Output:` as in previous examples because then the package runs this as a test and verifies that the output is exactly what we expect.

Reply:
Would you be against making this a file example, so I can make this a package level function?

Reply:
I had totally forgotten about whole file examples; seems reasonable to me to give it a go.